{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "assumed-opinion",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import statsmodels.api as sm\n",
    "from statsmodels.stats.mediation import Mediation\n",
    "import numpy as np\n",
    "import statsmodels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "directed-channels",
   "metadata": {},
   "outputs": [],
   "source": [
    "DEL_genotype = pd.read_csv('../preprocess_data/pangenie_svs_26sample_ALL_DEL_atleast5_50_genotype_only01.bed', sep='\\t', header=0, dtype=str)\n",
    "INS_genotype = pd.read_csv('../preprocess_data/pangenie_svs_26sample_ALL_INS_atleast5_50_genotype_only01.bed', sep='\\t', header=0, dtype=str)\n",
    "DEL_genotype = DEL_genotype.copy()\n",
    "INS_genotype = INS_genotype.copy()\n",
    "DEL_boundary_genotype = DEL_genotype[[\n",
    "'CHROM', 'START_POS', 'END_POS', 'TYPE', 'LEN', 'FILTER', 'INFO_ID','FORMAT','NA18534','NA18939','NA19036','NA19240','NA19650','NA19983','NA20509','NA20847','HG00096','HG00171','HG00514','HG00733','HG00864','HG01114','HG01505','HG01573','HG01596','HG02011','HG02018','HG02492','HG02587','HG03009','HG03065','HG03371','HG03683','HG03732'\n",
    "    ]]\n",
    "INS_boundary_genotype = INS_genotype[[\n",
    "'CHROM', 'START_POS', 'END_POS', 'TYPE', 'LEN', 'FILTER', 'INFO_ID','FORMAT','NA18534','NA18939','NA19036','NA19240','NA19650','NA19983','NA20509','NA20847','HG00096','HG00171','HG00514','HG00733','HG00864','HG01114','HG01505','HG01573','HG01596','HG02011','HG02018','HG02492','HG02587','HG03009','HG03065','HG03371','HG03683','HG03732'\n",
    "    ]]\n",
    "all_DEL_boundary_3_sig = pd.read_csv('../data/all_DEL_boundary_3_sig_10kbflank.bed', sep='\\t', header = 0)\n",
    "all_INS_boundary_3_sig = pd.read_csv('../data/all_INS_boundary_3_sig_10kbflank.bed', sep='\\t', header = 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "spare-while",
   "metadata": {},
   "source": [
    "## TAD-SV-eqtl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "certified-aquatic",
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_del_eqtl = pd.read_csv('../data/sig_del_eqtl.bed', sep = '\\t')\n",
    "sig_ins_eqtl = pd.read_csv('../data/sig_ins_eqtl.bed', sep = '\\t')\n",
    "sig_del_eqtl_new = sig_del_eqtl[['gene', 'INFO_ID', 'NA18534','NA18939','NA19036','NA19240','NA19650','NA19983','NA20509','NA20847','HG00096','HG00171','HG00514','HG00733','HG00864','HG01114','HG01505','HG01573','HG01596','HG02011','HG02018','HG02492','HG02587','HG03009','HG03065','HG03371','HG03683','HG03732']]\n",
    "sig_ins_eqtl_new = sig_ins_eqtl[['gene', 'INFO_ID', 'NA18534','NA18939','NA19036','NA19240','NA19650','NA19983','NA20509','NA20847','HG00096','HG00171','HG00514','HG00733','HG00864','HG01114','HG01505','HG01573','HG01596','HG02011','HG02018','HG02492','HG02587','HG03009','HG03065','HG03371','HG03683','HG03732']]\n",
    "sig_DEL_boundary_genotype = DEL_boundary_genotype[DEL_boundary_genotype['INFO_ID'].isin(sig_del_eqtl['INFO_ID'])]\n",
    "sig_DEL_boundary_genotype_new = sig_DEL_boundary_genotype[['INFO_ID', 'NA18534','NA18939','NA19036','NA19240','NA19650','NA19983','NA20509','NA20847','HG00096','HG00171','HG00514','HG00733','HG00864','HG01114','HG01505','HG01573','HG01596','HG02011','HG02018','HG02492','HG02587','HG03009','HG03065','HG03371','HG03683','HG03732']]\n",
    "sig_INS_boundary_genotype = INS_boundary_genotype[INS_boundary_genotype['INFO_ID'].isin(sig_ins_eqtl['INFO_ID'])]\n",
    "sig_INS_boundary_genotype_new = sig_INS_boundary_genotype[['INFO_ID', 'NA18534','NA18939','NA19036','NA19240','NA19650','NA19983','NA20509','NA20847','HG00096','HG00171','HG00514','HG00733','HG00864','HG01114','HG01505','HG01573','HG01596','HG02011','HG02018','HG02492','HG02587','HG03009','HG03065','HG03371','HG03683','HG03732']]\n",
    "sig_DEL_boundary_3 = all_DEL_boundary_3_sig[all_DEL_boundary_3_sig['INFO_ID'].isin(sig_del_eqtl['INFO_ID'])]\n",
    "sig_DEL_boundary_3_new = sig_DEL_boundary_3[['INFO_ID', 'GM18534','GM18939','GM19036','GM19240','GM19650','GM19833','GM20509','GM20847','HG00096','HG00171','HG00514','HG00733','HG00864','HG01114','HG01505','HG01573','HG01596','HG02011','HG02018','HG02492','HG02587','HG03009','HG03065','HG03371','HG03683','HG03732']]\n",
    "sig_INS_boundary_3 = all_INS_boundary_3_sig[all_INS_boundary_3_sig['INFO_ID'].isin(sig_ins_eqtl['INFO_ID'])]\n",
    "sig_INS_boundary_3_new = sig_INS_boundary_3[['INFO_ID', 'GM18534','GM18939','GM19036','GM19240','GM19650','GM19833','GM20509','GM20847','HG00096','HG00171','HG00514','HG00733','HG00864','HG01114','HG01505','HG01573','HG01596','HG02011','HG02018','HG02492','HG02587','HG03009','HG03065','HG03371','HG03683','HG03732']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "secondary-passion",
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_del_eqtl_new = sig_del_eqtl_new.copy()\n",
    "for index, row in sig_del_eqtl_new.iterrows():\n",
    "    gene = row['gene']\n",
    "    info_id = row['INFO_ID']\n",
    "    \n",
    "    # Extract rows from other datasets using INFO_ID\n",
    "    boundary_row = sig_DEL_boundary_3_new[sig_DEL_boundary_3_new['INFO_ID'] == info_id]\n",
    "    genotype_row = sig_DEL_boundary_genotype_new[sig_DEL_boundary_genotype_new['INFO_ID'] == info_id]\n",
    "     \n",
    "    # Check if the rows for the current INFO_ID are available in both datasets\n",
    "    if boundary_row.empty or genotype_row.empty:\n",
    "        continue\n",
    "    \n",
    "    # Prepare the data for analysis\n",
    "    Gene_expression = row.drop(labels=['INFO_ID', 'gene']).values.flatten()\n",
    "    Bound_strength = boundary_row.drop(columns=['INFO_ID']).values.flatten()\n",
    "    SV = genotype_row.drop(columns=['INFO_ID']).values.flatten()\n",
    "    \n",
    "    # Create a DataFrame for the current gene and INFO_ID\n",
    "    df = pd.DataFrame({\n",
    "        'Gene_expression': Gene_expression,\n",
    "        'Bound_strength': Bound_strength,\n",
    "        'SV': SV\n",
    "    })\n",
    "    \n",
    "    # Convert entire DataFrame to numeric, setting non-numeric values to NaN\n",
    "    df = df.apply(pd.to_numeric, errors='coerce')\n",
    "    \n",
    "    # Drop rows with NaN values (if any) due to type conversion\n",
    "    df = df.dropna()\n",
    "    \n",
    "    # Ensure that there are no missing values\n",
    "    if df.empty:\n",
    "        continue\n",
    "    \n",
    "    # Set a random seed for reproducibility\n",
    "    np.random.seed(710)\n",
    "\n",
    "    # The mediator model aims to examine how the independent variable (SV) affects the mediator variable (boundary strength).\n",
    "    mediator_formula = 'Bound_strength ~ SV'\n",
    "    # The outcome model examines how the independent variable (SV) and the mediator variable (boundary strength) together affect the dependent variable (gene expression).\n",
    "    outcome_formula = 'Gene_expression ~ SV + Bound_strength'\n",
    "\n",
    "    # Fit the mediator and outcome models using formulas\n",
    "    mediator_model = sm.OLS.from_formula(mediator_formula, df)\n",
    "    outcome_model = sm.OLS.from_formula(outcome_formula, df)\n",
    "    \n",
    "    # Perform mediation analysis\n",
    "    mediation_analysis = Mediation(outcome_model, mediator_model, 'SV', 'Bound_strength').fit()\n",
    "    \n",
    "    # Extract p-values from the mediation analysis summary\n",
    "    mediation_summary = mediation_analysis.summary()\n",
    "    pval_ACME = mediation_summary.loc['ACME (average)', 'P-value']\n",
    "    pval_ADE = mediation_summary.loc['ADE (average)', 'P-value']\n",
    "    pval_Total = mediation_summary.loc['Total effect', 'P-value']\n",
    "    \n",
    "    # Store the p-values in the sig_del_eqtl_new DataFrame\n",
    "    sig_del_eqtl_new.at[index, 'pval_ACME'] = pval_ACME\n",
    "    sig_del_eqtl_new.at[index, 'pval_ADE'] = pval_ADE\n",
    "    sig_del_eqtl_new.at[index, 'pval_Total'] = pval_Total"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "literary-fluid",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>INFO_ID</th>\n",
       "      <th>NA18534</th>\n",
       "      <th>NA18939</th>\n",
       "      <th>NA19036</th>\n",
       "      <th>NA19240</th>\n",
       "      <th>NA19650</th>\n",
       "      <th>NA19983</th>\n",
       "      <th>NA20509</th>\n",
       "      <th>NA20847</th>\n",
       "      <th>...</th>\n",
       "      <th>HG02492</th>\n",
       "      <th>HG02587</th>\n",
       "      <th>HG03009</th>\n",
       "      <th>HG03065</th>\n",
       "      <th>HG03371</th>\n",
       "      <th>HG03683</th>\n",
       "      <th>HG03732</th>\n",
       "      <th>pval_ACME</th>\n",
       "      <th>pval_ADE</th>\n",
       "      <th>pval_Total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZNF502</td>\n",
       "      <td>chr3-44510750-DEL-50</td>\n",
       "      <td>0.505437</td>\n",
       "      <td>0.711072</td>\n",
       "      <td>1.169339</td>\n",
       "      <td>0.680864</td>\n",
       "      <td>0.703322</td>\n",
       "      <td>0.658127</td>\n",
       "      <td>0.429553</td>\n",
       "      <td>0.838574</td>\n",
       "      <td>...</td>\n",
       "      <td>0.825207</td>\n",
       "      <td>0.906216</td>\n",
       "      <td>1.278297</td>\n",
       "      <td>0.618792</td>\n",
       "      <td>0.891358</td>\n",
       "      <td>1.553150</td>\n",
       "      <td>0.709216</td>\n",
       "      <td>0.514</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LOC105377056</td>\n",
       "      <td>chr3-44510750-DEL-50</td>\n",
       "      <td>0.051585</td>\n",
       "      <td>0.259984</td>\n",
       "      <td>0.248820</td>\n",
       "      <td>0.200549</td>\n",
       "      <td>0.133223</td>\n",
       "      <td>0.102748</td>\n",
       "      <td>0.153295</td>\n",
       "      <td>0.211812</td>\n",
       "      <td>...</td>\n",
       "      <td>0.189359</td>\n",
       "      <td>0.349621</td>\n",
       "      <td>0.228916</td>\n",
       "      <td>0.091698</td>\n",
       "      <td>0.161873</td>\n",
       "      <td>0.339314</td>\n",
       "      <td>0.184628</td>\n",
       "      <td>0.314</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XKR9</td>\n",
       "      <td>chr8-70671321-DEL-1088</td>\n",
       "      <td>0.230725</td>\n",
       "      <td>0.078901</td>\n",
       "      <td>0.301779</td>\n",
       "      <td>0.267848</td>\n",
       "      <td>0.270861</td>\n",
       "      <td>0.136379</td>\n",
       "      <td>0.530737</td>\n",
       "      <td>0.218365</td>\n",
       "      <td>...</td>\n",
       "      <td>0.225182</td>\n",
       "      <td>0.273766</td>\n",
       "      <td>0.139141</td>\n",
       "      <td>0.287291</td>\n",
       "      <td>0.175466</td>\n",
       "      <td>0.463090</td>\n",
       "      <td>0.130936</td>\n",
       "      <td>0.338</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERICH1</td>\n",
       "      <td>chr8-644401-DEL-5014</td>\n",
       "      <td>1.748027</td>\n",
       "      <td>2.397384</td>\n",
       "      <td>1.672566</td>\n",
       "      <td>2.079122</td>\n",
       "      <td>1.959391</td>\n",
       "      <td>1.973457</td>\n",
       "      <td>2.233838</td>\n",
       "      <td>2.596048</td>\n",
       "      <td>...</td>\n",
       "      <td>1.850300</td>\n",
       "      <td>1.798068</td>\n",
       "      <td>2.061475</td>\n",
       "      <td>1.680383</td>\n",
       "      <td>1.861537</td>\n",
       "      <td>2.319171</td>\n",
       "      <td>2.234540</td>\n",
       "      <td>0.904</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene                 INFO_ID   NA18534   NA18939   NA19036  \\\n",
       "0        ZNF502    chr3-44510750-DEL-50  0.505437  0.711072  1.169339   \n",
       "1  LOC105377056    chr3-44510750-DEL-50  0.051585  0.259984  0.248820   \n",
       "2          XKR9  chr8-70671321-DEL-1088  0.230725  0.078901  0.301779   \n",
       "3        ERICH1    chr8-644401-DEL-5014  1.748027  2.397384  1.672566   \n",
       "\n",
       "    NA19240   NA19650   NA19983   NA20509   NA20847  ...   HG02492   HG02587  \\\n",
       "0  0.680864  0.703322  0.658127  0.429553  0.838574  ...  0.825207  0.906216   \n",
       "1  0.200549  0.133223  0.102748  0.153295  0.211812  ...  0.189359  0.349621   \n",
       "2  0.267848  0.270861  0.136379  0.530737  0.218365  ...  0.225182  0.273766   \n",
       "3  2.079122  1.959391  1.973457  2.233838  2.596048  ...  1.850300  1.798068   \n",
       "\n",
       "    HG03009   HG03065   HG03371   HG03683   HG03732  pval_ACME  pval_ADE  \\\n",
       "0  1.278297  0.618792  0.891358  1.553150  0.709216      0.514     0.002   \n",
       "1  0.228916  0.091698  0.161873  0.339314  0.184628      0.314     0.000   \n",
       "2  0.139141  0.287291  0.175466  0.463090  0.130936      0.338     0.000   \n",
       "3  2.061475  1.680383  1.861537  2.319171  2.234540      0.904     0.000   \n",
       "\n",
       "   pval_Total  \n",
       "0       0.000  \n",
       "1       0.000  \n",
       "2       0.002  \n",
       "3       0.002  \n",
       "\n",
       "[4 rows x 31 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_del_eqtl_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "worldwide-nicaragua",
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_del_eqtl_new.to_csv('../data/all_del_eqtl_mediation.bed', index = False, sep='\\t', header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "disturbed-australia",
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_ins_eqtl_new = sig_ins_eqtl_new.copy()\n",
    "for index, row in sig_ins_eqtl_new.iterrows():\n",
    "    gene = row['gene']\n",
    "    info_id = row['INFO_ID']\n",
    "    \n",
    "    # Extract rows from other datasets using INFO_ID\n",
    "    boundary_row = sig_INS_boundary_3_new[sig_INS_boundary_3_new['INFO_ID'] == info_id]\n",
    "    genotype_row = sig_INS_boundary_genotype_new[sig_INS_boundary_genotype_new['INFO_ID'] == info_id]\n",
    "     \n",
    "    # Check if the rows for the current INFO_ID are available in both datasets\n",
    "    if boundary_row.empty or genotype_row.empty:\n",
    "        continue\n",
    "    \n",
    "    # Prepare the data for analysis\n",
    "    Gene_expression = row.drop(labels=['INFO_ID', 'gene']).values.flatten()\n",
    "    Bound_strength = boundary_row.drop(columns=['INFO_ID']).values.flatten()\n",
    "    SV = genotype_row.drop(columns=['INFO_ID']).values.flatten()\n",
    "    \n",
    "    # Create a DataFrame for the current gene and INFO_ID\n",
    "    df = pd.DataFrame({\n",
    "        'Gene_expression': Gene_expression,\n",
    "        'Bound_strength': Bound_strength,\n",
    "        'SV': SV\n",
    "    })\n",
    "    \n",
    "    # Convert entire DataFrame to numeric, setting non-numeric values to NaN\n",
    "    df = df.apply(pd.to_numeric, errors='coerce')\n",
    "    \n",
    "    # Drop rows with NaN values (if any) due to type conversion\n",
    "    df = df.dropna()\n",
    "    \n",
    "    # Ensure that there are no missing values\n",
    "    if df.empty:\n",
    "        continue\n",
    "    \n",
    "    # Set a random seed for reproducibility\n",
    "    np.random.seed(710)\n",
    "\n",
    "    # The mediator model aims to examine how the independent variable (SV) affects the mediator variable (boundary strength).\n",
    "    mediator_formula = 'Bound_strength ~ SV'\n",
    "    # The outcome model examines how the independent variable (SV) and the mediator variable (boundary strength) together affect the dependent variable (gene expression).\n",
    "    outcome_formula = 'Gene_expression ~ SV + Bound_strength'\n",
    "\n",
    "    # Fit the mediator and outcome models using formulas\n",
    "    mediator_model = sm.OLS.from_formula(mediator_formula, df)\n",
    "    outcome_model = sm.OLS.from_formula(outcome_formula, df)\n",
    "    \n",
    "    # Perform mediation analysis\n",
    "    mediation_analysis = Mediation(outcome_model, mediator_model, 'SV', 'Bound_strength').fit()\n",
    "    \n",
    "    # Extract p-values from the mediation analysis summary\n",
    "    mediation_summary = mediation_analysis.summary()\n",
    "    pval_ACME = mediation_summary.loc['ACME (average)', 'P-value']\n",
    "    pval_ADE = mediation_summary.loc['ADE (average)', 'P-value']\n",
    "    pval_Total = mediation_summary.loc['Total effect', 'P-value']\n",
    "    \n",
    "    # Store the p-values in the sig_del_eqtl_new DataFrame\n",
    "    sig_ins_eqtl_new.at[index, 'pval_ACME'] = pval_ACME\n",
    "    sig_ins_eqtl_new.at[index, 'pval_ADE'] = pval_ADE\n",
    "    sig_ins_eqtl_new.at[index, 'pval_Total'] = pval_Total"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "simplified-saudi",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>INFO_ID</th>\n",
       "      <th>NA18534</th>\n",
       "      <th>NA18939</th>\n",
       "      <th>NA19036</th>\n",
       "      <th>NA19240</th>\n",
       "      <th>NA19650</th>\n",
       "      <th>NA19983</th>\n",
       "      <th>NA20509</th>\n",
       "      <th>NA20847</th>\n",
       "      <th>...</th>\n",
       "      <th>HG02492</th>\n",
       "      <th>HG02587</th>\n",
       "      <th>HG03009</th>\n",
       "      <th>HG03065</th>\n",
       "      <th>HG03371</th>\n",
       "      <th>HG03683</th>\n",
       "      <th>HG03732</th>\n",
       "      <th>pval_ACME</th>\n",
       "      <th>pval_ADE</th>\n",
       "      <th>pval_Total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCDC163</td>\n",
       "      <td>chr1-45497763-INS-354</td>\n",
       "      <td>2.451951</td>\n",
       "      <td>2.138003</td>\n",
       "      <td>2.807306</td>\n",
       "      <td>1.613811</td>\n",
       "      <td>2.671173</td>\n",
       "      <td>2.302246</td>\n",
       "      <td>2.088805</td>\n",
       "      <td>2.481762</td>\n",
       "      <td>...</td>\n",
       "      <td>2.596948</td>\n",
       "      <td>2.678364</td>\n",
       "      <td>2.705544</td>\n",
       "      <td>1.964997</td>\n",
       "      <td>2.047240</td>\n",
       "      <td>2.476094</td>\n",
       "      <td>2.718013</td>\n",
       "      <td>0.248</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SCAPER</td>\n",
       "      <td>chr15-76198866-INS-67</td>\n",
       "      <td>2.732734</td>\n",
       "      <td>2.450514</td>\n",
       "      <td>2.412903</td>\n",
       "      <td>2.561219</td>\n",
       "      <td>2.473855</td>\n",
       "      <td>2.413278</td>\n",
       "      <td>2.501987</td>\n",
       "      <td>2.962800</td>\n",
       "      <td>...</td>\n",
       "      <td>1.970033</td>\n",
       "      <td>2.210502</td>\n",
       "      <td>2.369374</td>\n",
       "      <td>2.151918</td>\n",
       "      <td>1.861709</td>\n",
       "      <td>2.555223</td>\n",
       "      <td>2.630280</td>\n",
       "      <td>0.086</td>\n",
       "      <td>0.226</td>\n",
       "      <td>0.024</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene                INFO_ID   NA18534   NA18939   NA19036   NA19240  \\\n",
       "0  CCDC163  chr1-45497763-INS-354  2.451951  2.138003  2.807306  1.613811   \n",
       "1   SCAPER  chr15-76198866-INS-67  2.732734  2.450514  2.412903  2.561219   \n",
       "\n",
       "    NA19650   NA19983   NA20509   NA20847  ...   HG02492   HG02587   HG03009  \\\n",
       "0  2.671173  2.302246  2.088805  2.481762  ...  2.596948  2.678364  2.705544   \n",
       "1  2.473855  2.413278  2.501987  2.962800  ...  1.970033  2.210502  2.369374   \n",
       "\n",
       "    HG03065   HG03371   HG03683   HG03732  pval_ACME  pval_ADE  pval_Total  \n",
       "0  1.964997  2.047240  2.476094  2.718013      0.248     0.002       0.022  \n",
       "1  2.151918  1.861709  2.555223  2.630280      0.086     0.226       0.024  \n",
       "\n",
       "[2 rows x 31 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_ins_eqtl_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "interested-animal",
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_ins_eqtl_new.to_csv('../data/all_ins_eqtl_mediation.bed', index = False, sep='\\t', header=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "appreciated-figure",
   "metadata": {},
   "source": [
    "## TAD-SV-sqtl "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "other-zoning",
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_del_sqtl = pd.read_csv('../data/sig_del_sqtl.bed', sep = '\\t')\n",
    "sig_ins_sqtl = pd.read_csv('../data/sig_ins_sqtl.bed', sep = '\\t')\n",
    "sig_del_sqtl_new = sig_del_sqtl[['feature_id', 'INFO_ID', 'NA18534','NA18939','NA19036','NA19240','NA19650','NA19983','NA20509','NA20847','HG00096','HG00171','HG00514','HG00733','HG00864','HG01114','HG01505','HG01573','HG01596','HG02011','HG02018','HG02492','HG02587','HG03009','HG03065','HG03371','HG03683','HG03732']]\n",
    "sig_ins_sqtl_new = sig_ins_sqtl[['feature_id', 'INFO_ID', 'NA18534','NA18939','NA19036','NA19240','NA19650','NA19983','NA20509','NA20847','HG00096','HG00171','HG00514','HG00733','HG00864','HG01114','HG01505','HG01573','HG01596','HG02011','HG02018','HG02492','HG02587','HG03009','HG03065','HG03371','HG03683','HG03732']]\n",
    "sig_DEL_boundary_genotype = DEL_boundary_genotype[DEL_boundary_genotype['INFO_ID'].isin(sig_del_sqtl['INFO_ID'])]\n",
    "sig_INS_boundary_genotype = INS_boundary_genotype[INS_boundary_genotype['INFO_ID'].isin(sig_ins_sqtl['INFO_ID'])]\n",
    "sig_DEL_boundary_genotype_new = sig_DEL_boundary_genotype[['INFO_ID', 'NA18534','NA18939','NA19036','NA19240','NA19650','NA19983','NA20509','NA20847','HG00096','HG00171','HG00514','HG00733','HG00864','HG01114','HG01505','HG01573','HG01596','HG02011','HG02018','HG02492','HG02587','HG03009','HG03065','HG03371','HG03683','HG03732']]\n",
    "sig_INS_boundary_genotype_new = sig_INS_boundary_genotype[['INFO_ID', 'NA18534','NA18939','NA19036','NA19240','NA19650','NA19983','NA20509','NA20847','HG00096','HG00171','HG00514','HG00733','HG00864','HG01114','HG01505','HG01573','HG01596','HG02011','HG02018','HG02492','HG02587','HG03009','HG03065','HG03371','HG03683','HG03732']]\n",
    "sig_DEL_boundary_3 = all_DEL_boundary_3_sig[all_DEL_boundary_3_sig['INFO_ID'].isin(sig_del_sqtl['INFO_ID'])]\n",
    "sig_DEL_boundary_3_new = sig_DEL_boundary_3[['INFO_ID', 'GM18534','GM18939','GM19036','GM19240','GM19650','GM19833','GM20509','GM20847','HG00096','HG00171','HG00514','HG00733','HG00864','HG01114','HG01505','HG01573','HG01596','HG02011','HG02018','HG02492','HG02587','HG03009','HG03065','HG03371','HG03683','HG03732']]\n",
    "sig_INS_boundary_3 = all_INS_boundary_3_sig[all_INS_boundary_3_sig['INFO_ID'].isin(sig_ins_sqtl['INFO_ID'])]\n",
    "sig_INS_boundary_3_new = sig_INS_boundary_3[['INFO_ID', 'GM18534','GM18939','GM19036','GM19240','GM19650','GM19833','GM20509','GM20847','HG00096','HG00171','HG00514','HG00733','HG00864','HG01114','HG01505','HG01573','HG01596','HG02011','HG02018','HG02492','HG02587','HG03009','HG03065','HG03371','HG03683','HG03732']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "respiratory-weekly",
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_del_sqtl_new = sig_del_sqtl_new.copy()\n",
    "for index, row in sig_del_sqtl_new.iterrows():\n",
    "    feature_id = row['feature_id']\n",
    "    info_id = row['INFO_ID']\n",
    "    \n",
    "    # Extract rows from other datasets using INFO_ID\n",
    "    boundary_row = sig_DEL_boundary_3_new[sig_DEL_boundary_3_new['INFO_ID'] == info_id]\n",
    "    genotype_row = sig_DEL_boundary_genotype_new[sig_DEL_boundary_genotype_new['INFO_ID'] == info_id]\n",
    "     \n",
    "    # Check if the rows for the current INFO_ID are available in both datasets\n",
    "    if boundary_row.empty or genotype_row.empty:\n",
    "        continue\n",
    "    \n",
    "    # Prepare the data for analysis\n",
    "    Splicing_ratio = row.drop(labels=['INFO_ID', 'feature_id']).values.flatten()\n",
    "    Bound_strength = boundary_row.drop(columns=['INFO_ID']).values.flatten()\n",
    "    SV = genotype_row.drop(columns=['INFO_ID']).values.flatten()\n",
    "    \n",
    "    # Create a DataFrame for the current feature_id and INFO_ID\n",
    "    df = pd.DataFrame({\n",
    "        'Splicing_ratio': Splicing_ratio,\n",
    "        'Bound_strength': Bound_strength,\n",
    "        'SV': SV\n",
    "    })\n",
    "    \n",
    "    # Convert entire DataFrame to numeric, setting non-numeric values to NaN\n",
    "    df = df.apply(pd.to_numeric, errors='coerce')\n",
    "    \n",
    "    # Drop rows with NaN values (if any) due to type conversion\n",
    "    df = df.dropna()\n",
    "    \n",
    "    # Ensure that there are no missing values\n",
    "    if df.empty:\n",
    "        continue\n",
    "    \n",
    "    # Set a random seed for reproducibility\n",
    "    np.random.seed(710)\n",
    "\n",
    "    # The mediator model aims to examine how the independent variable (SV) affects the mediator variable (boundary strength).\n",
    "    mediator_formula = 'Bound_strength ~ SV'\n",
    "    # The outcome model examines how the independent variable (SV) and the mediator variable (boundary strength) together affect the dependent variable (Splicing_ratio).\n",
    "    outcome_formula = 'Splicing_ratio ~ SV + Bound_strength'\n",
    "\n",
    "    # Fit the mediator and outcome models using formulas\n",
    "    mediator_model = sm.OLS.from_formula(mediator_formula, df)\n",
    "    outcome_model = sm.OLS.from_formula(outcome_formula, df)\n",
    "    \n",
    "    # Perform mediation analysis\n",
    "    mediation_analysis = Mediation(outcome_model, mediator_model, 'SV', 'Bound_strength').fit()\n",
    "    \n",
    "    # Extract p-values from the mediation analysis summary\n",
    "    mediation_summary = mediation_analysis.summary()\n",
    "    pval_ACME = mediation_summary.loc['ACME (average)', 'P-value']\n",
    "    pval_ADE = mediation_summary.loc['ADE (average)', 'P-value']\n",
    "    pval_Total = mediation_summary.loc['Total effect', 'P-value']\n",
    "    \n",
    "    # Store the p-values in the sig_del_sqtl_new DataFrame\n",
    "    sig_del_sqtl_new.at[index, 'pval_ACME'] = pval_ACME\n",
    "    sig_del_sqtl_new.at[index, 'pval_ADE'] = pval_ADE\n",
    "    sig_del_sqtl_new.at[index, 'pval_Total'] = pval_Total"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "computational-thompson",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_id</th>\n",
       "      <th>INFO_ID</th>\n",
       "      <th>NA18534</th>\n",
       "      <th>NA18939</th>\n",
       "      <th>NA19036</th>\n",
       "      <th>NA19240</th>\n",
       "      <th>NA19650</th>\n",
       "      <th>NA19983</th>\n",
       "      <th>NA20509</th>\n",
       "      <th>NA20847</th>\n",
       "      <th>...</th>\n",
       "      <th>HG02492</th>\n",
       "      <th>HG02587</th>\n",
       "      <th>HG03009</th>\n",
       "      <th>HG03065</th>\n",
       "      <th>HG03371</th>\n",
       "      <th>HG03683</th>\n",
       "      <th>HG03732</th>\n",
       "      <th>pval_ACME</th>\n",
       "      <th>pval_ADE</th>\n",
       "      <th>pval_Total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6:87391835:87392725:clu_24429_+</td>\n",
       "      <td>chr6-87408456-DEL-133</td>\n",
       "      <td>-0.936184</td>\n",
       "      <td>0.012348</td>\n",
       "      <td>1.076310</td>\n",
       "      <td>2.158304</td>\n",
       "      <td>-0.932265</td>\n",
       "      <td>0.412867</td>\n",
       "      <td>-0.841126</td>\n",
       "      <td>-0.767154</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.040756</td>\n",
       "      <td>-1.141381</td>\n",
       "      <td>0.385166</td>\n",
       "      <td>-0.996222</td>\n",
       "      <td>-1.159694</td>\n",
       "      <td>0.681233</td>\n",
       "      <td>-0.554325</td>\n",
       "      <td>0.736</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8:70669538:70674818:clu_22301_+</td>\n",
       "      <td>chr8-70671321-DEL-1088</td>\n",
       "      <td>-0.005980</td>\n",
       "      <td>-0.950419</td>\n",
       "      <td>-0.399550</td>\n",
       "      <td>-0.929802</td>\n",
       "      <td>2.013419</td>\n",
       "      <td>-1.227758</td>\n",
       "      <td>2.069432</td>\n",
       "      <td>-1.077835</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.646946</td>\n",
       "      <td>1.773509</td>\n",
       "      <td>-1.187269</td>\n",
       "      <td>0.009941</td>\n",
       "      <td>1.334757</td>\n",
       "      <td>0.612355</td>\n",
       "      <td>-0.937157</td>\n",
       "      <td>0.498</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8:664676:668598:clu_8652_-</td>\n",
       "      <td>chr8-644401-DEL-5014</td>\n",
       "      <td>1.245457</td>\n",
       "      <td>-0.414052</td>\n",
       "      <td>1.217779</td>\n",
       "      <td>-0.759694</td>\n",
       "      <td>1.209294</td>\n",
       "      <td>1.131286</td>\n",
       "      <td>-0.186955</td>\n",
       "      <td>0.427311</td>\n",
       "      <td>...</td>\n",
       "      <td>1.370163</td>\n",
       "      <td>1.423127</td>\n",
       "      <td>0.506337</td>\n",
       "      <td>1.393303</td>\n",
       "      <td>0.331152</td>\n",
       "      <td>0.336981</td>\n",
       "      <td>0.661554</td>\n",
       "      <td>0.304</td>\n",
       "      <td>0.222</td>\n",
       "      <td>0.026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8:70652752:70653695:clu_22301_+</td>\n",
       "      <td>chr8-70671321-DEL-1088</td>\n",
       "      <td>0.704935</td>\n",
       "      <td>1.305387</td>\n",
       "      <td>0.883035</td>\n",
       "      <td>0.859624</td>\n",
       "      <td>-1.808690</td>\n",
       "      <td>0.580523</td>\n",
       "      <td>-2.483752</td>\n",
       "      <td>0.948607</td>\n",
       "      <td>...</td>\n",
       "      <td>0.560184</td>\n",
       "      <td>-1.196977</td>\n",
       "      <td>0.807123</td>\n",
       "      <td>-0.060676</td>\n",
       "      <td>-0.594006</td>\n",
       "      <td>0.057426</td>\n",
       "      <td>0.818711</td>\n",
       "      <td>0.884</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>20:35632923:35664760:clu_21372_-</td>\n",
       "      <td>chr20-35960724-DEL-60</td>\n",
       "      <td>1.746169</td>\n",
       "      <td>0.529841</td>\n",
       "      <td>0.298889</td>\n",
       "      <td>-0.216541</td>\n",
       "      <td>-0.391229</td>\n",
       "      <td>0.191113</td>\n",
       "      <td>0.014453</td>\n",
       "      <td>-0.368097</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.028003</td>\n",
       "      <td>-0.167166</td>\n",
       "      <td>-0.169622</td>\n",
       "      <td>-0.432385</td>\n",
       "      <td>-0.916697</td>\n",
       "      <td>-0.475933</td>\n",
       "      <td>-0.963485</td>\n",
       "      <td>0.786</td>\n",
       "      <td>0.024</td>\n",
       "      <td>0.024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>20:35655344:35658930:clu_21372_-</td>\n",
       "      <td>chr20-35960724-DEL-60</td>\n",
       "      <td>-2.666742</td>\n",
       "      <td>-0.113100</td>\n",
       "      <td>-0.047915</td>\n",
       "      <td>0.324691</td>\n",
       "      <td>0.682168</td>\n",
       "      <td>0.111165</td>\n",
       "      <td>0.232633</td>\n",
       "      <td>0.199351</td>\n",
       "      <td>...</td>\n",
       "      <td>0.629328</td>\n",
       "      <td>0.722698</td>\n",
       "      <td>0.518111</td>\n",
       "      <td>0.397488</td>\n",
       "      <td>0.963883</td>\n",
       "      <td>0.527364</td>\n",
       "      <td>1.150455</td>\n",
       "      <td>0.796</td>\n",
       "      <td>0.120</td>\n",
       "      <td>0.084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>20:35632923:35658930:clu_21372_-</td>\n",
       "      <td>chr20-35960724-DEL-60</td>\n",
       "      <td>0.152098</td>\n",
       "      <td>0.418184</td>\n",
       "      <td>0.353582</td>\n",
       "      <td>0.031710</td>\n",
       "      <td>-0.009831</td>\n",
       "      <td>1.008811</td>\n",
       "      <td>0.817291</td>\n",
       "      <td>-0.411666</td>\n",
       "      <td>...</td>\n",
       "      <td>0.239788</td>\n",
       "      <td>-0.015342</td>\n",
       "      <td>0.578768</td>\n",
       "      <td>-0.050955</td>\n",
       "      <td>-2.056413</td>\n",
       "      <td>-0.244741</td>\n",
       "      <td>-1.089664</td>\n",
       "      <td>0.948</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>20:35626381:35626735:clu_21369_-</td>\n",
       "      <td>chr20-35960724-DEL-60</td>\n",
       "      <td>-0.432567</td>\n",
       "      <td>0.036543</td>\n",
       "      <td>-0.758824</td>\n",
       "      <td>0.380615</td>\n",
       "      <td>0.448366</td>\n",
       "      <td>-0.411455</td>\n",
       "      <td>0.123351</td>\n",
       "      <td>0.158921</td>\n",
       "      <td>...</td>\n",
       "      <td>0.846607</td>\n",
       "      <td>0.066323</td>\n",
       "      <td>0.321752</td>\n",
       "      <td>0.041438</td>\n",
       "      <td>0.769644</td>\n",
       "      <td>0.496257</td>\n",
       "      <td>0.992429</td>\n",
       "      <td>0.552</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>20:35626381:35626567:clu_21369_-</td>\n",
       "      <td>chr20-35960724-DEL-60</td>\n",
       "      <td>0.548644</td>\n",
       "      <td>0.026216</td>\n",
       "      <td>0.522271</td>\n",
       "      <td>-0.070876</td>\n",
       "      <td>-0.411626</td>\n",
       "      <td>0.380335</td>\n",
       "      <td>-0.271860</td>\n",
       "      <td>-0.091484</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.176643</td>\n",
       "      <td>0.101298</td>\n",
       "      <td>-0.041055</td>\n",
       "      <td>-0.358616</td>\n",
       "      <td>-0.959188</td>\n",
       "      <td>0.294405</td>\n",
       "      <td>-1.346577</td>\n",
       "      <td>0.660</td>\n",
       "      <td>0.012</td>\n",
       "      <td>0.008</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         feature_id                 INFO_ID   NA18534  \\\n",
       "0   6:87391835:87392725:clu_24429_+   chr6-87408456-DEL-133 -0.936184   \n",
       "1   8:70669538:70674818:clu_22301_+  chr8-70671321-DEL-1088 -0.005980   \n",
       "2        8:664676:668598:clu_8652_-    chr8-644401-DEL-5014  1.245457   \n",
       "3   8:70652752:70653695:clu_22301_+  chr8-70671321-DEL-1088  0.704935   \n",
       "4  20:35632923:35664760:clu_21372_-   chr20-35960724-DEL-60  1.746169   \n",
       "5  20:35655344:35658930:clu_21372_-   chr20-35960724-DEL-60 -2.666742   \n",
       "6  20:35632923:35658930:clu_21372_-   chr20-35960724-DEL-60  0.152098   \n",
       "7  20:35626381:35626735:clu_21369_-   chr20-35960724-DEL-60 -0.432567   \n",
       "8  20:35626381:35626567:clu_21369_-   chr20-35960724-DEL-60  0.548644   \n",
       "\n",
       "    NA18939   NA19036   NA19240   NA19650   NA19983   NA20509   NA20847  ...  \\\n",
       "0  0.012348  1.076310  2.158304 -0.932265  0.412867 -0.841126 -0.767154  ...   \n",
       "1 -0.950419 -0.399550 -0.929802  2.013419 -1.227758  2.069432 -1.077835  ...   \n",
       "2 -0.414052  1.217779 -0.759694  1.209294  1.131286 -0.186955  0.427311  ...   \n",
       "3  1.305387  0.883035  0.859624 -1.808690  0.580523 -2.483752  0.948607  ...   \n",
       "4  0.529841  0.298889 -0.216541 -0.391229  0.191113  0.014453 -0.368097  ...   \n",
       "5 -0.113100 -0.047915  0.324691  0.682168  0.111165  0.232633  0.199351  ...   \n",
       "6  0.418184  0.353582  0.031710 -0.009831  1.008811  0.817291 -0.411666  ...   \n",
       "7  0.036543 -0.758824  0.380615  0.448366 -0.411455  0.123351  0.158921  ...   \n",
       "8  0.026216  0.522271 -0.070876 -0.411626  0.380335 -0.271860 -0.091484  ...   \n",
       "\n",
       "    HG02492   HG02587   HG03009   HG03065   HG03371   HG03683   HG03732  \\\n",
       "0 -1.040756 -1.141381  0.385166 -0.996222 -1.159694  0.681233 -0.554325   \n",
       "1 -0.646946  1.773509 -1.187269  0.009941  1.334757  0.612355 -0.937157   \n",
       "2  1.370163  1.423127  0.506337  1.393303  0.331152  0.336981  0.661554   \n",
       "3  0.560184 -1.196977  0.807123 -0.060676 -0.594006  0.057426  0.818711   \n",
       "4 -0.028003 -0.167166 -0.169622 -0.432385 -0.916697 -0.475933 -0.963485   \n",
       "5  0.629328  0.722698  0.518111  0.397488  0.963883  0.527364  1.150455   \n",
       "6  0.239788 -0.015342  0.578768 -0.050955 -2.056413 -0.244741 -1.089664   \n",
       "7  0.846607  0.066323  0.321752  0.041438  0.769644  0.496257  0.992429   \n",
       "8 -0.176643  0.101298 -0.041055 -0.358616 -0.959188  0.294405 -1.346577   \n",
       "\n",
       "   pval_ACME  pval_ADE  pval_Total  \n",
       "0      0.736     0.008       0.002  \n",
       "1      0.498     0.000       0.000  \n",
       "2      0.304     0.222       0.026  \n",
       "3      0.884     0.000       0.000  \n",
       "4      0.786     0.024       0.024  \n",
       "5      0.796     0.120       0.084  \n",
       "6      0.948     0.008       0.014  \n",
       "7      0.552     0.026       0.012  \n",
       "8      0.660     0.012       0.008  \n",
       "\n",
       "[9 rows x 31 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_del_sqtl_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "awful-marathon",
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_del_sqtl_new.to_csv('../data/all_del_sqtl_mediation.bed', index = False, sep='\\t', header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "confused-washer",
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_ins_sqtl_new = sig_ins_sqtl_new.copy()\n",
    "for index, row in sig_ins_sqtl_new.iterrows():\n",
    "    feature_id = row['feature_id']\n",
    "    info_id = row['INFO_ID']\n",
    "    \n",
    "    # Extract rows from other datasets using INFO_ID\n",
    "    boundary_row = sig_INS_boundary_3_new[sig_INS_boundary_3_new['INFO_ID'] == info_id]\n",
    "    genotype_row = sig_INS_boundary_genotype_new[sig_INS_boundary_genotype_new['INFO_ID'] == info_id]\n",
    "     \n",
    "    # Check if the rows for the current INFO_ID are available in both datasets\n",
    "    if boundary_row.empty or genotype_row.empty:\n",
    "        continue\n",
    "    \n",
    "    # Prepare the data for analysis\n",
    "    Splicing_ratio = row.drop(labels=['INFO_ID', 'feature_id']).values.flatten()\n",
    "    Bound_strength = boundary_row.drop(columns=['INFO_ID']).values.flatten()\n",
    "    SV = genotype_row.drop(columns=['INFO_ID']).values.flatten()\n",
    "    \n",
    "    # Create a DataFrame for the current feature_id and INFO_ID\n",
    "    df = pd.DataFrame({\n",
    "        'Splicing_ratio': Splicing_ratio,\n",
    "        'Bound_strength': Bound_strength,\n",
    "        'SV': SV\n",
    "    })\n",
    "    \n",
    "    # Convert entire DataFrame to numeric, setting non-numeric values to NaN\n",
    "    df = df.apply(pd.to_numeric, errors='coerce')\n",
    "    \n",
    "    # Drop rows with NaN values (if any) due to type conversion\n",
    "    df = df.dropna()\n",
    "    \n",
    "    # Ensure that there are no missing values\n",
    "    if df.empty:\n",
    "        continue\n",
    "    \n",
    "    # Set a random seed for reproducibility\n",
    "    np.random.seed(710)\n",
    "\n",
    "    # The mediator model aims to examine how the independent variable (SV) affects the mediator variable (TAD strength or boundary strength).\n",
    "    mediator_formula = 'Bound_strength ~ SV'\n",
    "    # The outcome model examines how the independent variable (SV) and the mediator variable (boundary strength) together affect the dependent variable (Splicing_ratio).\n",
    "    outcome_formula = 'Splicing_ratio ~ SV + Bound_strength'\n",
    "\n",
    "    # Fit the mediator and outcome models using formulas\n",
    "    mediator_model = sm.OLS.from_formula(mediator_formula, df)\n",
    "    outcome_model = sm.OLS.from_formula(outcome_formula, df)\n",
    "    \n",
    "    # Perform mediation analysis\n",
    "    mediation_analysis = Mediation(outcome_model, mediator_model, 'SV', 'Bound_strength').fit()\n",
    "    \n",
    "    # Extract p-values from the mediation analysis summary\n",
    "    mediation_summary = mediation_analysis.summary()\n",
    "    pval_ACME = mediation_summary.loc['ACME (average)', 'P-value']\n",
    "    pval_ADE = mediation_summary.loc['ADE (average)', 'P-value']\n",
    "    pval_Total = mediation_summary.loc['Total effect', 'P-value']\n",
    "    \n",
    "    # Store the p-values in the sig_ins_sqtl_new DataFrame\n",
    "    sig_ins_sqtl_new.at[index, 'pval_ACME'] = pval_ACME\n",
    "    sig_ins_sqtl_new.at[index, 'pval_ADE'] = pval_ADE\n",
    "    sig_ins_sqtl_new.at[index, 'pval_Total'] = pval_Total"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "exciting-sleeping",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_id</th>\n",
       "      <th>INFO_ID</th>\n",
       "      <th>NA18534</th>\n",
       "      <th>NA18939</th>\n",
       "      <th>NA19036</th>\n",
       "      <th>NA19240</th>\n",
       "      <th>NA19650</th>\n",
       "      <th>NA19983</th>\n",
       "      <th>NA20509</th>\n",
       "      <th>NA20847</th>\n",
       "      <th>...</th>\n",
       "      <th>HG02492</th>\n",
       "      <th>HG02587</th>\n",
       "      <th>HG03009</th>\n",
       "      <th>HG03065</th>\n",
       "      <th>HG03371</th>\n",
       "      <th>HG03683</th>\n",
       "      <th>HG03732</th>\n",
       "      <th>pval_ACME</th>\n",
       "      <th>pval_ADE</th>\n",
       "      <th>pval_Total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:45551155:45552436:clu_13870_+</td>\n",
       "      <td>chr1-45497763-INS-354</td>\n",
       "      <td>-1.746480</td>\n",
       "      <td>0.475494</td>\n",
       "      <td>-0.078211</td>\n",
       "      <td>-0.445241</td>\n",
       "      <td>-0.698239</td>\n",
       "      <td>-2.578916</td>\n",
       "      <td>0.338158</td>\n",
       "      <td>1.013390</td>\n",
       "      <td>...</td>\n",
       "      <td>0.861487</td>\n",
       "      <td>0.718653</td>\n",
       "      <td>-0.177377</td>\n",
       "      <td>-0.359558</td>\n",
       "      <td>-0.607451</td>\n",
       "      <td>0.442949</td>\n",
       "      <td>-0.515916</td>\n",
       "      <td>0.452</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19:37507047:37508265:clu_27679_+</td>\n",
       "      <td>chr19-37789443-INS-1092</td>\n",
       "      <td>-0.422717</td>\n",
       "      <td>0.996525</td>\n",
       "      <td>1.166925</td>\n",
       "      <td>1.740052</td>\n",
       "      <td>-1.491070</td>\n",
       "      <td>1.274219</td>\n",
       "      <td>-1.437013</td>\n",
       "      <td>-1.299856</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.421464</td>\n",
       "      <td>-1.307926</td>\n",
       "      <td>-1.419196</td>\n",
       "      <td>-0.705460</td>\n",
       "      <td>2.317070</td>\n",
       "      <td>-1.440027</td>\n",
       "      <td>2.429939</td>\n",
       "      <td>0.936</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.008</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         feature_id                  INFO_ID   NA18534  \\\n",
       "0   1:45551155:45552436:clu_13870_+    chr1-45497763-INS-354 -1.746480   \n",
       "1  19:37507047:37508265:clu_27679_+  chr19-37789443-INS-1092 -0.422717   \n",
       "\n",
       "    NA18939   NA19036   NA19240   NA19650   NA19983   NA20509   NA20847  ...  \\\n",
       "0  0.475494 -0.078211 -0.445241 -0.698239 -2.578916  0.338158  1.013390  ...   \n",
       "1  0.996525  1.166925  1.740052 -1.491070  1.274219 -1.437013 -1.299856  ...   \n",
       "\n",
       "    HG02492   HG02587   HG03009   HG03065   HG03371   HG03683   HG03732  \\\n",
       "0  0.861487  0.718653 -0.177377 -0.359558 -0.607451  0.442949 -0.515916   \n",
       "1 -1.421464 -1.307926 -1.419196 -0.705460  2.317070 -1.440027  2.429939   \n",
       "\n",
       "   pval_ACME  pval_ADE  pval_Total  \n",
       "0      0.452     0.002       0.002  \n",
       "1      0.936     0.014       0.008  \n",
       "\n",
       "[2 rows x 31 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_ins_sqtl_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "jewish-boundary",
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_ins_sqtl_new.to_csv('../data/all_ins_sqtl_mediation.bed', index = False, sep='\\t', header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "earlier-footage",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
